mergify_engine/merge_fast_forwarder.py

```diff
@@ -0,0 +1,124 @@
+# -*- encoding: utf-8 -*-
+#
+# Copyright © 2022 Mergify SAS
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+import dataclasses
+import typing
+
+from mergify_engine import exceptions
+from mergify_engine import gitter
+
+
+if typing.TYPE_CHECKING:
+    from mergify_engine import context
+    from mergify_engine.dashboard import user_tokens
+
+
+@dataclasses.dataclass
+class MergeFastForwardUnexpectedError(Exception):
+    reason: str
+
+
+@dataclasses.dataclass
+class MergeFastForwardNeedRetry(exceptions.EngineNeedRetry):
+    reason: str
+
+
+@dataclasses.dataclass
+class MergeFastForwardFailed(Exception):
+    reason: str
+
+
+@dataclasses.dataclass
+class MergeFastForwardBaseMoved(Exception):
+    reason: str
+
+
+GIT_MESSAGE_TO_EXCEPTION = {
+    "(non-fast-forward)": MergeFastForwardFailed,
+    "Updates were rejected because the tip of your current branch is behind": MergeFastForwardBaseMoved,
+    "You may want to first integrate the remote changes": MergeFastForwardBaseMoved,
+    "couldn't find remote ref": MergeFastForwardFailed,
+}
+
```

Comment:
I hope we have tests for all those; changing the git version would break the tests in that case and it would be good to know.

Reply:
We don't have any tests that reproduce git errors, I pick those from branch_updater and duplicate_pull